_shared/project/README.md

```diff
@@ -0,0 +1,263 @@
+
+```terminal
+pipx upgrade {{ cookiecutter.slug }}
+```
+
+To see what version you have run:
+
+```terminal
+{{ cookiecutter.__entry_point }} --version
+```
+
+## Uninstalling
+
+To uninstall run:
+
+```
+pipx uninstall {{ cookiecutter.slug }}
+```
+{%- endif %}
+
+## Setting up Your {{ cookiecutter.name }} Development Environment
+
+First you'll need to install:
+
+* [Git](https://git-scm.com/).
+  On Ubuntu: `sudo apt install git`, on macOS: `brew install git`.
+* [GNU Make](https://www.gnu.org/software/make/).
+  This is probably already installed, run `make --version` to check.
+* [pyenv](https://github.com/pyenv/pyenv).
+  Follow the instructions in pyenv's README to install it.
+  The **Homebrew** method works best on macOS.
+  The **Basic GitHub Checkout** method works best on Ubuntu.
+  You _don't_ need to set up pyenv's shell integration ("shims"), you can
+  [use pyenv without shims](https://github.com/pyenv/pyenv#using-pyenv-without-shims).
+{%- if cookiecutter.get("services") == "yes" %}
+* [Docker](https://docs.docker.com/install/).
+  Follow the [instructions on the Docker website](https://docs.docker.com/install/)
+  to install it.  
+  You **don't** need to install Docker Compose, the development environment
+  will install it automatically for you in tox.  
+  You **do** need to set up the `docker` command to work without `sudo`,
+  on Linux this means following Docker's [Post-installation steps for Linux](https://docs.docker.com/engine/install/linux-postinstall/).
+{%- endif %}
+{%- if cookiecutter.get("frontend") == "yes" %}
+* [Node](https://nodejs.org/) and npm.
+  On Ubuntu: `sudo snap install --classic node`.
+  On macOS: `brew install node`.
+* [Yarn](https://yarnpkg.com/): `sudo npm install -g yarn`.
+{%- endif %}
+{{- include("hacking/prerequisites") }}
+
+Then to set up your development environment:
+
+```terminal
+git clone {{ cookiecutter.__github_url }}.git
+cd {{ cookiecutter.package_name }}
+{%- if cookiecutter.get("services") == "yes" %}
+make services
+{%- endif %}
+{%- if cookiecutter.get("devdata") == "yes" %}
+make devdata
+{%- endif %}
+make help
+```
+
+{%- if cookiecutter._directory == 'pypackage' %}
+
+## Releasing a New Version of the Project
+
+1. First, to get PyPI publishing working you need to go to:
+   <https://github.com/organizations/hypothesis/settings/secrets/actions/PYPI_TOKEN>
+   and add {{ cookiecutter.slug }} to the `PYPI_TOKEN` secret's selected
+   repositories.
+
+2. Now that the {{ cookiecutter.slug }} project has access to the `PYPI_TOKEN` secret
+   you can release a new version by just [creating a new GitHub release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository).
+   Publishing a new GitHub release will automatically trigger
+   [a GitHub Actions workflow](.github/workflows/pypi.yml)
+   that will build the new version of your Python package and upload it to
+   <{{ cookiecutter.__pypi_url }}>.
+{% endif -%}
+
```

Comment:
These blocks are so long it's quite hard to see what the ifs and else's correlate to.

I think we might be better off with a separate include blocks here which have the parts for the various projects in separate files. Then you could have something like:

```jinja2
{% if cookiecutter._directory == 'pypackage' %}
    {{ include("hacking/pypackage/release") }}
    {{ include("hacking/pypackage/python_versions") }}
{% elif cookiecutter._directory == 'pyramid-app' %}
    {{ include("hacking/pyramid_app/python_versions") }}
{% endif %}
```

Reply:
Another thought is if we have this much which is custom to one type of project or another, but is not "shared" in a meaningful way, is this the right place for it?

We could replace this with `{% block custom_content %}` and use `extend` to keep the parts which are specific to pypackages close to the templates they serve.

Reply:
Responded to this in the [long comment below](https://github.com/hypothesis/cookiecutters/pull/79#issuecomment-1307220125). TLDR: I don't think we can split the `README.md` template into separate files: we can use `include`, `extend` or `block`. But I think breaking it up into macros still within a single file might help with this.

I think macros would allow us to have a concise "map" of the logic of the file similar to the example you posted:

```jinja2
{% if cookiecutter._directory == 'pypackage' %}
    {{ release() }}
    {{ package)python_versions() }}
{% elif cookiecutter._directory == 'pyramid-app' %}
    {{ app_python_versions() }}
{% endif %}
```

It's just that all the macros need to be defined at the top of the `README.md` file (they can't be defined in other files) and then the logical map would be at the bottom of the file. See the longer comment below for details.

Reply:
See https://github.com/hypothesis/cookiecutters/pull/79#issuecomment-1307238767 for the mix of shared and specific content